### ISSUE TITLE
Wrong order of Insert then Update during batch saveAll

### ISSUE THREAD

**Issue by plinyar** [ebean-orm/ebean#3193]
Imagine, I have an entity Tariff with fields: (id, countryId, pecentage, isDefault) and the business rule, that in a country there must be not more than 1 default tariff. In order to guarantee consistency, I create in database a unique constraint so that no 2 tuples (countryId, default or null) should exist.

So if user inserts new Tariff and wants it to become default, my application logic should update old default tariff (reset default flag) and insert the new one.

The logic manipulates with both instances and executes Database.saveAll(the2instances) to save them in a single batch.

## Expected behavior
eBean should first execute entities to update, and only then to insert

## Actual behavior
Now eBean first inserts the entities of a batch, then updates. This will always cause unique constraint violations even if the batch containt updates to avoid the violation.

## Comment
Why saveAll? In our application we tend to aggregate all changes in business layer to a Unit of Work, and them apply the Unit of Work through eBean's batching mechanism. Now we have to overcome the sequence of execution in a custom code smth like (Kotlin):

```
fun <E:Any> Database.run(unitOfWork: UnitOfWork<E>) {
    deleteAll(unitOfWork.toDelete)

//   This fails:
//    saveAll(unitOfWork.toPersist) - toPersists returns single list of entities in the order: to update, to save (upsert), to insert

// So instead we do:
    val (toUpdate, toInsert) = unitOfWork.toUpsert.partition { isUpdate(it) }
    updateAll(unitOfWork.toUpdate + toUpdate)
    insertAll(unitOfWork.toUpsert + toInsert)
}

private fun interceptOf(entity:Any) = (entity as EntityBean)._ebean_getIntercept()
private fun isUpdate(entity:Any) = interceptOf(entity).isUpdate

```

**Comment by rbygrave** [comment]
My expectation is that instead of using `saveAll()` you would instead use `@Transaction(batchSize=...)` or an explicit transaction and turn on batch mode.  Then call `save()` in the order that is required.

**Comment by rbygrave** [comment]
In BatchedBeanHolder that holds the beans of the same type (and level) in insert, update and delete batches.

They are persisted in the order of:
- deletes
- inserts
- updates

I believe you are asking to change that order to be deletes, updates, inserts.

**Comment by plinyar** [comment]
> I believe you are asking to change that order to be deletes, updates, inserts.

Indeed. I believe it should be:
deletes
updates
inserts

**Comment by jnehlmeier** [comment]
As long as you use saveAll() the same issue could appear if you have two existing instances and they should swap their default boolean. The javadoc of saveAll() does not say anything about query ordering compared to collection ordering. So if you have two modified instances it could very well happen that the first update query will set the boolean to true which would already trigger the unique constraint unless you configure the unique constraint to be deferred and be validated at commit time or you take control in the application and execute queries in correct ordering. 

I don't think Ebean will generate a single sql update that toggles the boolean for both modified rows to make the update atomic. But even if it would, it would be an implementation detail. Currently the saveAll() behavior is also an implementation detail.

What happens if the batch ordering would be changed and someone has modified existing entities to reference new entities? Obviously inserts would need to be done first in order to execute the updates and satisfy foreign keys. Would this change possibly break existing apps?

**Comment by rbygrave** [comment]
> What happens if the batch ordering would be changed and someone has modified existing entities to reference new entities? Obviously inserts would need to be done first in order to execute the updates and satisfy foreign keys. Would this change possibly break existing apps?

This is the concern. A self referencing table that desires the insert to be first, would break with a change to the ordering as mooted here. 

Deferring the constraint to be enforced at commit would not require an ordering change.

**Comment by plinyar** [comment]
@jnehlmeier Jens, thanks for the point with the self-referenced table. Now it is obvious that order of inserts/updates may have to be different for the different cases, and as @rbygrave Rob mentioned, the deferred constraints will be the solution.

It is reasonable to assume that saveAll guarantees no specific order of inserts/updates.

Unfortunately, for my particular case (it is Postgres) I found no way to make the constraint deferred by default (since the unique index is partial, with filter to exclude rows with isDefault = false), and I will have to manipulate with 'deferred' things at transaction level.
